### DISCUSSION TITLE
How to apply the same SQL filtering code to queries with different `FromClause`s

### DISCUSSION THREAD

**Discussion by davidatsurge** [diesel-rs/diesel#3468] [Q&A] [answered]
With schema and imports:

```rust
use diesel::prelude::*;
use diesel::{pg::Pg, query_dsl::methods::BoxedDsl, dsl::count_star};

mod schema {
    diesel::table! {
        ingredients (id) {
            id -> Uuid,
            name -> Text,
        }
    }
    diesel::table! {
        recipe_ingredients (ingredient_id, recipe_id) {
            ingredient_id -> Uuid,
            recipe_id -> Uuid,
        }
    }
    diesel::table! {
            recipes (id) {
                id -> Uuid,
                aggregate_rating_value -> Nullable<Float4>,
        }
    }

    diesel::joinable!(recipe_ingredients -> ingredients (ingredient_id));
    diesel::joinable!(recipe_ingredients -> recipes (recipe_id));

    diesel::allow_tables_to_appear_in_same_query!(
        ingredients,
        recipe_ingredients,
        recipes,
    );
}
```

I want to apply the same SQL filtering (as defined in a `filter_recipes` function) to queries with different `FromClause`s.

```rust
struct GetRecipesParams {
    min_avg_rating: Option<f32>,
}

fn main() {
    let params = GetRecipesParams {
        min_avg_rating: Some(4.),
    };

    let fetched_recipes_count = filter_recipes(
        schema::recipes::table.select(count_star()),
        &params
    );

    let fetched_recipes = filter_recipes(
        schema::recipes::table
            .left_join(schema::recipe_ingredients::table.inner_join(schema::ingredients::table))
            .select(recipes::all_columns),
        &params
    );
}
```

But I can't figure out what the trait bounds on the arguments of `filter_recipes` should be. Here's my (wrong) attempt:

```rust
fn filter_recipes<'a, T>(
    recipes_query: T,
    params: &GetRecipesParams,
) -> <T as BoxedDsl<'a, diesel::pg::Pg>>::Output
where
    T: diesel::Table,
    T: BoxedDsl<'a, Pg>,
{
    use schema::recipes;
    let GetRecipesParams { min_avg_rating } = params;

    let recipes_query = recipes_query.into_boxed();

    if let Some(min_avg_rating) = min_avg_rating {
        recipes_query = db::diesel::query_dsl::methods::FilterDsl::filter(
            recipes_query,
            recipes::aggregate_rating_value.ge(min_avg_rating),
        )
    };

    recipes_query
}
```

**Comment by weiznich** [comment, marked as answer]
The easiest solution is to just apply the filter before the join/any other operation. Otherwise quite complex generic trait bounds are required.

For reference the following code works for me:

```rust
fn main() {
    let params = GetRecipesParams {
        min_avg_rating: Some(4.),
    };

    let fetched_recipes_count = filter_recipes(schema::recipes::table.into_boxed(), &params)
        .select(diesel::dsl::count_star());

    let fetched_recipes = filter_recipes(schema::recipes::table.into_boxed(), &params)
        .left_join(schema::recipe_ingredients::table.inner_join(schema::ingredients::table))
        .select(schema::recipes::all_columns);
}

fn filter_recipes<'a>(
    mut recipes_query: schema::recipes::BoxedQuery<'a, diesel::pg::Pg>,
    params: &'a GetRecipesParams,
) -> schema::recipes::BoxedQuery<'a, diesel::pg::Pg> {
    use schema::recipes;
    let GetRecipesParams { min_avg_rating } = params;

    if let Some(min_avg_rating) = min_avg_rating {
        recipes_query = recipes_query.filter(recipes::aggregate_rating_value.ge(min_avg_rating))
    };

    recipes_query
}

```

**Reply by davidatsurge** [reply]
EDIT: Georg, please read the next reply that I sent first. I'm having trouble articulating myself, tbh, and I say what I _actually_ want in the next reply.

Thank you very much for the response Georg! Ok, it turns out I wasn't precise / articulate about what I actually wanted. What I actually want is to apply the same filtering to two queries, one with a GROUP BY and one without.

So, I tried to work off of the solution you posted above and where I am is:

```rust
fn main() {
    let params = GetRecipesParams {
        min_avg_rating: Some(4.),
    };

    let fetched_recipes_count = filter_recipes(schema::recipes::table.into_boxed(), &params)
        .select(schema::recipes::id)
        .distinct()
        .count();

    let fetched_recipes = filter_recipes(
        schema::recipes::table
            .group_by(schema::recipes::all_columns)
            .into_boxed(),
        &params,
    )
    .select(schema::recipes::all_columns)
    .left_join(schema::recipe_ingredients::table.inner_join(schema::ingredients::table)); // The left-join errors out complaining that `T: Table` is not satisfied.
}

fn filter_recipes<'a, ST, GB>(
    mut recipes_query: diesel::internal::table_macro::BoxedSelectStatement<
        'a,
        ST,
        ::diesel::internal::table_macro::FromClause<schema::recipes::table>,
        diesel::pg::Pg,
        GB,
    >,
    params: &'a GetRecipesParams,
) -> diesel::query_builder::BoxedSelectStatement<
    'a,
    ST,
    diesel::query_builder::FromClause<schema::recipes::table>,
    diesel::pg::Pg,
    GB,
> {
    use schema::recipes;
    let GetRecipesParams { min_avg_rating } = params;

    if let Some(min_avg_rating) = min_avg_rating {
        recipes_query = recipes_query.filter(recipes::aggregate_rating_value.ge(min_avg_rating))
    };

    recipes_query
}
```

And in addition to that, I also want to be able to do filtering based on the JOINed columns in the `filter_recipes` function, (which I haven't tried to address in my attempts above).

**Reply by davidatsurge** [reply]
Ok, let me be even more clear.

What I _actually_ want is just to get the count of rows  in a statement with a GROUP BY. But the issue is, if I just do `.count()` it will count the number in each group, which I don't want. I want the total number of grouped rows. What I want to achieve in SQL is:

```sql
SELECT COUNT(*) FROM (SELECT id FROM recipes GROUP BY id);
```

But I couldn't figure out how to have a select statement in a FROM clause in diesel, which is why I thought "I know! To get the count, I'll avoid the GROUP BY and do `.select(grouping_cols).distinct().count()` instead, which is what spawned this whole thread.

But again, what I TRULY want is summarized in that SQL excerpt above.

**Reply by weiznich** [reply]
Isn't that query equivalent to `SELECT count(other_colum) FROM recipes GROUP BY id`? In that case you can write it just as `recipes::table.group_by(recipes::id).select(dsl::count(recipes::id))`

Other than that: Selecting from a subselect in that way is not supported by diesel yet. You need to write your own dsl extension for that or just use `sql_query` if you truly want to express exactly that query.

**Reply by davidatsurge** [reply]
Well in this case, `id` is guaranteed to be unique for each row, so yes, the two would be equivalent, but for a column that is not guaranteed to be unique, they would not be equivalent, which is what I really wanted. It seems like I'm on a roll when it comes to _not_ expressing what I really want.

In any case, I got around this by instead doing `recipes::table.select(recipes::id.eq_any(some_complex_query.select(recipes::id))).count()`, which is exactly what I wanted.

I honestly want to thank you so much for engaging me in this thread. Especially given that I was incredibly inarticulate for some reason!
